src/beatree/writeout.rs

```diff
@@ -264,18 +281,62 @@ impl BbnWriteOut {
 struct LnWriteOut {
     ln_fd: RawFd,
     ln_extend_file_sz: Option<u64>,
-    ln: Vec<()>,
+    ln: Vec<(PageNumber, Box<Page>)>,
+    ln_remaining: usize,
 }
 
 impl LnWriteOut {
-    fn run(&mut self) -> bool {
+    fn run(&mut self, io: &mut IoDmux) -> bool {
+        // NOTE: a possible improvement could be to allow the LeafStore to perform write operations
+        // directly into itself if it falls within an already allocated file.
+        // The `writeout` function would then only need to call ftruncate and write any exceeded pages.
+
         if let Some(new_len) = self.ln_extend_file_sz.take() {
             unsafe {
                 File::from_raw_fd(self.ln_fd).set_len(new_len).unwrap();
             }
         }
 
-        // TODO: fill this out similar to BBN write out.
+        while self.ln.len() > 0 {
+            // UNWRAP: we know the list is not empty.
+            let (ln_pn, ln_page) = self.ln.pop().unwrap();
+
+            {
+                // UNWRAP: the branch node cannot be out of bounds because of the requirement of the
+                // sync machine.
+                // TODO: can this clone be avoided? maybe from Box to Rc? or using WriteRaw
```

Comment:
we definitely need to up our game wrt buffer management, but isn't it the case that you get your message back and you could extract the page from there to put it back?

Reply:
you're totally right! thanks!!